File: AutoRun/AutoRun.py
import os
import sys
import time

import BubbleSort
import HeapSort
import InsertionSort
import MergeSort
import QuickSort
import SelectionSort

#Generate Log Folders
main_dir = os.path.dirname(os.path.abspath(sys.argv[0]))
log_dir = main_dir + "\Logs"
if (os.path.exists(log_dir) == False):
    os.makedirs(log_dir)
algo_dirs = ["\Bubble", "\Heap", "\Insertion", "\Merge", "\Quick", "\Selection"]
for i in algo_dirs:
        if (os.path.exists(log_dir + i) == False):
            os.makedirs(log_dir + i)

def file2array(filename):
    array = []
    file = open(filename, "r")
    for i in file:
        array = i.split(",")
    return array

def compute(filename, sort_type, pass_num):
    data = file2array(filename)
    time_before = time.time()
    if (sort_type == "Bubble"):
        BubbleSort.BubbleSort(data)
        tempdir = log_dir + "\Bubble"
    if (sort_type == "Heap"):
        HeapSort.HeapSort(data)
        tempdir = log_dir + "\Heap"
    if (sort_type == "Insertion"):
        InsertionSort.InsertionSort(data)
        tempdir = log_dir + "\Insertion"
    if (sort_type == "Merge"):
        MergeSort.MergeSort(data)
        tempdir = log_dir + "\Merge"
    if (sort_type == "Quick"):
        QuickSort.QuickSort(data, 0, len(data) - 1)
        tempdir = log_dir + "\Quick"
    if (sort_type == "Selection"):
        SelectionSort.SelectionSort(data, len(data) - 1)
        tempdir = log_dir + "\Selection"
    time_after = time.time()
    time_diff = time_after - time_before
    logname = tempdir + "\Log_" + sort_type + "Sort_" + str(pass_num) + ".txt"
    try:
        compute_log = open(logname, "a")
    except:
        compute_log = open(logname, "w")
    log = "Applying " + sort_type + "Sort to " + filename + "\nSorted Array:\n"
    log += "Time to sort: " + str(time_diff) + "\n"
    compute_log.write(log)
    compute_log.close

num_passes = 2
num_arrays = 2
algo_types = ["Bubble", "Heap", "Insertion", "Merge", "Quick", "Selection"]
for i in algo_types:
    j = 1
    while j <= num_passes:
        k = 1
        while k <= num_arrays:
            compute(main_dir + "\Arrays\Array_" + str(k) + ".txt", i, j)
            k += 1
        j += 1

File: AutoRun C++/AutoRun.cpp
#include <iostream>
#include <iomanip>
#include <string>
#include <fstream>
#include <sstream>
#include <chrono>
#include <filesystem>

#include "BubbleSort.cpp"
#include "HeapSort.cpp"
#include "InsertionSort.cpp"
#include "MergeSort.cpp"
#include "QuickSort.cpp"
#include "SelectionSort.cpp"

int getSize(string filename);
void file2array(string filename, int* &array);
void compute(string filename);


using namespace std; 
namespace fs = std::filesystem;


//gets the total number of numbers in the array file to be used in a dynamically allocated array
int getSize(string filename) {
    ifstream file(filename);
    string line; 
    getline(file, line);

    int count = 0;
    istringstream iss(line);
    string number;
    while (getline(iss, number, ',')) {
        count++;
    }
    file.close();

    return count;

}

//reads the array file and stores the numbers in a dynamically allocated array
void file2array(string filename, int* &array) {
  ifstream file(filename);
  string line; 
  getline(file, line);

  istringstream iss(line);
  string number;
  int count = 0;
  while (getline(iss, number, ',')) {
      array[count++] = stoi(number);
  }

  file.close();
}

//creates a new dynamically allocated array, inserts the data from the current array file, and measures the execution time of the current sort algorithm.
void compute(string filename, string sortType, string logDir, int pass) {
  int size = getSize(filename);
  int* array = new int[size];
  string tempDir;

  file2array(filename, array);

  //measures execution time in microseconds
  auto start = std::chrono::high_resolution_clock::now();
  if (sortType == "Bubble") {
    bubbleSort(array, size);
    tempDir = logDir + "\\Bubble";
  }
  if (sortType == "Heap") {
    heapSort(array, size);
    tempDir = logDir + "\\Heap";
  }
  if (sortType == "Insertion") {
    insertionSort(array, size);
    tempDir = logDir + "\\Insertion";
  }
  if (sortType == "Merge") {
    mergeSort(array, 0, size - 1);
    tempDir = logDir + "\\Merge";
  }
  if (sortType == "Quick") {
    quickSort(array, 0, size - 1);
    tempDir = logDir + "\\Quick";
  }
  if (sortType == "Selection") {
    selectionSort(array, size);
    tempDir = logDir + "\\Selection";
  }
  auto stop = std::chrono::high_resolution_clock::now();
  double duration = std::chrono::duration_cast<std::chrono::microseconds>(stop - start).count();

  //converts microseconds to seconds
  duration *= 1e-6;

  //creates a new log file
  string logName = tempDir + "\\Log_" + sortType + "Sort_" + to_string(pass) + ".txt";
  ofstream newLog(logName, std::ios::app);
  newLog << "Applying " << sortType << "Sort to " << filename << "\nTime to sort: " << fixed << setprecision(6) << duration << " seconds\n" << endl;
  newLog.close();
  delete[] array;
}

int main() 
{  
  //make "Logs" folder
  string mainDir = fs::current_path().string();
  string logDir = mainDir + "\\Logs";
  if (!fs::is_directory(logDir)) {
    fs::create_directory(logDir); 
  }
  //make each algorithm log folder
  string algoDirs[6] = {"\\Bubble", "\\Heap", "\\Insertion", "\\Merge", "\\Quick", "\\Selection"};
  for (int i = 0; i < 6; i++) {
    if (!fs::is_directory(logDir + algoDirs[i])) {
    fs::create_directory(logDir + algoDirs[i]); 
    }
  }

  string sortTypes[6] = {"Bubble", "Heap", "Insertion", "Merge", "Quick", "Selection"};


  //can edit numPasses and numArrays to create different number of logs and test different number of arrays
  int numPasses = 2;
  int numArrays = 10; //max numArrays = 10
  //clears any previous logs
  for (int type = 0; type < 6; type++) {
    for (int pass = 0; pass < numPasses; pass++) {
      for (int array = 0; array < numArrays; array++) {
        string logName = logDir + "\\" + sortTypes[type] + "\\Log_" + sortTypes[type] + "Sort_" + to_string(pass) + ".txt";
        ofstream clearLog(logName, std::ios::trunc);
      }
    }
  }

  string arrayDir = mainDir + "\\Arrays";
  for (int type = 0; type < 6; type++) {
    for (int pass = 0; pass < numPasses; pass++) {
      for (int array = 0; array < numArrays; array++) {
        compute(arrayDir + "\\Array_" + to_string(array + 1) + ".txt", sortTypes[type], logDir, pass);
      }
    }
  }

  return 0;
}